In [1]:
!if cd combinedmatrixlib; then git pull && cd ..; else \
git clone https://ivangilcruz@bitbucket.org/ivangilcruz/combinedmatrixlib.git; fi

/bin/sh: 1: cd: can't cd to combinedmatrixlib
Cloning into 'combinedmatrixlib'...
remote: Counting objects: 147, done.
remote: Compressing objects: 100% (144/144), done.
remote: Total 147 (delta 81), reused 0 (delta 0)
Receiving objects: 100% (147/147), 19.96 KiB | 1.05 MiB/s, done.
Resolving deltas: 100% (81/81), done.


In [0]:
from combinedmatrixlib.matrix import *

In [0]:
msa=Matrix.create_symmetric_matrix(4, arbitrary=False, aleatory=True, interval=[0,5], integerEntry=True)

In [0]:
msg=Matrix.create_symmetric_matrix(4, arbitrary=True, aleatory=True, interval=[0,5], integerEntry=True)

In [0]:
m1=Matrix.create_totally_positive_matrix(4, interval=[0,5], integerEntry=True)["product"]

In [0]:
mc=m1.combined_matrix()

In [7]:
mc

Matrix([
[  7895/2, -12843/2,   3150,     -675],
[-20943/2,    19285, -12375,   7125/2],
[   11025, -93375/4,  18928, -26433/4],
[   -4500,  41925/4,  -9702,  14887/4]])

In [8]:
sum(mc.row(3))

1

In [9]:
sum(mc.col(2))

1

In [10]:
for i in range(100):
    m1=Matrix.create_totally_positive_matrix(4, interval=[0,1000000], integerEntry=True, symmetric=True)['product']
    mc=m1.combined_matrix()
    d_mc=mc.get_diagonals()
    if (d_mc[1]+d_mc[2]>d_mc[3]+d_mc[0])==False:
        print "No se cumple"
        break
print "Fin"

Fin


In [0]:
tnp = Matrix.create_totally_nonpositive_matrix(4, interval=[1,5], integerEntry=True)["product"]

In [5]:
tnp

Matrix([
[-788/135, -788/45, -788/15, -788/5],
[ -788/27,  -752/9,  -692/3,   -612],
[-1576/27, -1432/9, -1186/3,   -854],
[-3152/27, -2648/9, -1760/3,   -540]])

$\left[\begin{matrix}- \frac{521}{150} & - \frac{521}{30} & - \frac{521}{15} & - \frac{2084}{15}\\- \frac{521}{50} & - \frac{481}{10} & - \frac{421}{5} & - \frac{1504}{5}\\- \frac{521}{10} & - \frac{457}{2} & -357 & -1116\\- \frac{1042}{5} & -854 & -1096 & -2400\end{matrix}\right]$


In [0]:
d=tnp.get_diagonals()

In [14]:
d

[-641/360, -569/24, -2341/8, -3600]

In [15]:
d[0] + d[2] > d[1]+d[3]-1

True

In [22]:
for i in range(100):
    m1=Matrix.create_totally_nonpositive_matrix(4, interval=[1,10], integerEntry=True, symmetric=True)['product']
    mc=m1.combined_matrix()
    d_mc=mc.get_diagonals()
    if (d_mc[0] + d_mc[2] >= d_mc[1]+d_mc[3]-1)==False:
        print "No se cumple"
        m1.to_pdf(title="Matriz tnp: tnp1", filename="tnp")
        mc.to_pdf(title="Matriz combinada de tnp1", filename="matriz_combinada")
        break
print "Fin"

No se cumple
Fin


In [17]:
for i in range(1000):
    m1=Matrix.create_totally_nonpositive_matrix(4, interval=[1,1000000], integerEntry=True, symmetric=True)['product']
    mc=m1.combined_matrix()
    d_mc=mc.get_diagonals()
    if ((d_mc[1] - d_mc[0]+1)*d_mc[3] > -1*d_mc[2])==False:
        print "No se cumple"
        m1.to_pdf(title="Matriz tnp: tnp2", filename="tnp2")
        mc.to_pdf(title="Matriz combinada de tnp2", filename="matriz_combinada_tnp2")
        break
print "Fin"

Fin


In [18]:
for i in range(1000):
    m1=Matrix.create_totally_nonpositive_matrix(4, interval=[1,1000], integerEntry=True, symmetric=True)['product']
    mc=m1.combined_matrix()
    d_mc=mc.get_diagonals()
    if (-1 > d_mc[3])==False:
        print "No se cumple"
        m1.to_pdf(title="Matriz tnp: tnp1", filename="tnp")
        mc.to_pdf(title="Matriz combinada de tnp1", filename="matriz_combinada")
        break
print "Fin"

Fin


In [20]:
for i in range(1000):
    m1=Matrix.create_totally_nonpositive_matrix(4, interval=[1,1000], integerEntry=True, symmetric=True)['product']
    mc=m1.combined_matrix()
    d_mc=mc.get_diagonals()
    if (-1 > d_mc[0])==False:
        print "No se cumple"
        m1.to_pdf(title="Matriz tnp: tnp1", filename="tnp")
        mc.to_pdf(title="Matriz combinada de tnp1", filename="matriz_combinada")
        break
print "Fin"

Fin


In [11]:
for i in range(1000):
    m1=Matrix.create_totally_nonpositive_matrix(4, interval=[1,1000], integerEntry=True, symmetric=True)['product']
    mc=m1.combined_matrix()
    d_mc=mc.get_diagonals()
    if (-1 > d_mc[2])==False:
        print "No se cumple"
        m1.to_pdf(title="Matriz tnp: tnp1", filename="tnp")
        mc.to_pdf(title="Matriz combinada de tnp1", filename="matriz_combinada")
        break
print "Fin"

Fin


In [6]:
for i in range(100):
    m1=Matrix.create_totally_nonpositive_matrix(3, interval=[1,1000], integerEntry=True, symmetric=True)['product']
    mc=m1.combined_matrix()
    d_mc=mc.get_diagonals()
    if False in [x <= -1 for x in d_mc]:
        print "No se cumple"
        #m1.to_pdf(title="Matriz tnp: tnp1", filename="tnp")
        #mc.to_pdf(title="Matriz combinada de tnp1", filename="matriz_combinada")
        break
print "Fin"

Fin


# Matrices bidiagonales superiores e inferiores con entradas symbolicas

In [11]:
#Lower bidiagonals
print "\n".join(["$"+sympy.latex(m)+"$" for m in Matrix.create_lower_bidiagonal_matrix(4, symbolic=True)])

$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & 0 & 1 & 0\\0 & 0 & a & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & c & 1 & 0\\0 & 0 & b & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\f & 1 & 0 & 0\\0 & e & 1 & 0\\0 & 0 & d & 1\end{matrix}\right]$


$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & 0 & 1 & 0\\0 & 0 & a & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & c & 1 & 0\\0 & 0 & b & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\f & 1 & 0 & 0\\0 & e & 1 & 0\\0 & 0 & d & 1\end{matrix}\right]$

In [8]:
#Diagonal
print "$"+sympy.latex(Matrix.create_diagonal_matrix(4, symbolic=True))+"$"

$\left[\begin{matrix}g & 0 & 0 & 0\\0 & h & 0 & 0\\0 & 0 & i & 0\\0 & 0 & 0 & j\end{matrix}\right]$


$\left[\begin{matrix}g & 0 & 0 & 0\\0 & h & 0 & 0\\0 & 0 & i & 0\\0 & 0 & 0 & j\end{matrix}\right]$

In [10]:
#Upper bidiagonals
print "\n".join(["$"+sympy.latex(m)+"$" for m in Matrix.create_upper_bidiagonal_matrix(4, symbolic=True)])

$\left[\begin{matrix}1 & k & 0 & 0\\0 & 1 & l & 0\\0 & 0 & 1 & m\\0 & 0 & 0 & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & n & 0\\0 & 0 & 1 & o\\0 & 0 & 0 & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & 0 & 1 & p\\0 & 0 & 0 & 1\end{matrix}\right]$


$\left[\begin{matrix}1 & k & 0 & 0\\0 & 1 & l & 0\\0 & 0 & 1 & m\\0 & 0 & 0 & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & n & 0\\0 & 0 & 1 & o\\0 & 0 & 0 & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & 0 & 1 & p\\0 & 0 & 0 & 1\end{matrix}\right]$

In [0]:
TP = Matrix.create_totally_positive_matrix(4,symbolic=True)["product"]

In [22]:
print "$" + sympy.latex(TP) + "$"

$\left[\begin{matrix}g & g k & g k n & g k n p\\f g & f g k + h & h l + n \left(f g k + h\right) & h l o + p \left(h l + n \left(f g k + h\right)\right)\\c f g & c f g k + h \left(c + e\right) & h l \left(c + e\right) + i + n \left(c f g k + h \left(c + e\right)\right) & i m + o \left(h l \left(c + e\right) + i\right) + p \left(h l \left(c + e\right) + i + n \left(c f g k + h \left(c + e\right)\right)\right)\\a c f g & a c f g k + h \left(a c + e \left(a + b\right)\right) & h l \left(a c + e \left(a + b\right)\right) + i \left(a + b + d\right) + n \left(a c f g k + h \left(a c + e \left(a + b\right)\right)\right) & i m \left(a + b + d\right) + j + o \left(h l \left(a c + e \left(a + b\right)\right) + i \left(a + b + d\right)\right) + p \left(h l \left(a c + e \left(a + b\right)\right) + i \left(a + b + d\right) + n \left(a c f g k + h \left(a c + e \left(a + b\right)\right)\right)\right)\end{matrix}\right]$


$\left[\begin{matrix}g & g k & g k n & g k n p\\f g & f g k + h & h l + n \left(f g k + h\right) & h l o + p \left(h l + n \left(f g k + h\right)\right)\\c f g & c f g k + h \left(c + e\right) & h l \left(c + e\right) + i + n \left(c f g k + h \left(c + e\right)\right) & i m + o \left(h l \left(c + e\right) + i\right) + p \left(h l \left(c + e\right) + i + n \left(c f g k + h \left(c + e\right)\right)\right)\\a c f g & a c f g k + h \left(a c + e \left(a + b\right)\right) & h l \left(a c + e \left(a + b\right)\right) + i \left(a + b + d\right) + n \left(a c f g k + h \left(a c + e \left(a + b\right)\right)\right) & i m \left(a + b + d\right) + j + o \left(h l \left(a c + e \left(a + b\right)\right) + i \left(a + b + d\right)\right) + p \left(h l \left(a c + e \left(a + b\right)\right) + i \left(a + b + d\right) + n \left(a c f g k + h \left(a c + e \left(a + b\right)\right)\right)\right)\end{matrix}\right]$

In [0]:
TP_inv = TP.inv().expand()

In [0]:
TP_inv_diag = TP_inv.get_diagonals()

In [13]:
print "\n".join(["$"+sympy.latex(d) + "$" for d in TP_inv_diag])

$\frac{d e}{j} f k l m + \frac{e f}{i} k l + \frac{f k}{h} + \frac{1}{g}$
$\frac{b c}{j} l m + \frac{b c}{j} m n + \frac{b c}{j} n o + \frac{c d}{j} l m + \frac{c d}{j} m n + \frac{c d}{j} n o + \frac{c l}{i} + \frac{c n}{i} + \frac{d e}{j} l m + \frac{d e}{j} m n + \frac{d e}{j} n o + \frac{e l}{i} + \frac{e n}{i} + \frac{1}{h}$
$\frac{a m}{j} + \frac{a o}{j} + \frac{a p}{j} + \frac{b m}{j} + \frac{b o}{j} + \frac{b p}{j} + \frac{d m}{j} + \frac{d o}{j} + \frac{d p}{j} + \frac{1}{i}$
$\frac{1}{j}$


$\frac{d e}{j} f k l m + \frac{e f}{i} k l + \frac{f k}{h} + \frac{1}{g}$

$\frac{b c}{j} l m + \frac{b c}{j} m n + \frac{b c}{j} n o + \frac{c d}{j} l m + \frac{c d}{j} m n + \frac{c d}{j} n o + \frac{c l}{i} + \frac{c n}{i} + \frac{d e}{j} l m + \frac{d e}{j} m n + \frac{d e}{j} n o + \frac{e l}{i} + \frac{e n}{i} + \frac{1}{h}$

$\frac{a m}{j} + \frac{a o}{j} + \frac{a p}{j} + \frac{b m}{j} + \frac{b o}{j} + \frac{b p}{j} + \frac{d m}{j} + \frac{d o}{j} + \frac{d p}{j} + \frac{1}{i}$

$\frac{1}{j}$

# Matriz totalmente no positiva con entradas simbolicas

In [25]:
#Lower bidiagonals
print "\n".join(["$"+sympy.latex(m)+"$" for m in Matrix.create_lower_bidiagonal_matrix(4, symbolic=True)])

$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & 0 & 1 & 0\\0 & 0 & a & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & c & 1 & 0\\0 & 0 & b & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\f & 1 & 0 & 0\\0 & e & 1 & 0\\0 & 0 & d & 1\end{matrix}\right]$


$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & 0 & 1 & 0\\0 & 0 & a & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & c & 1 & 0\\0 & 0 & b & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\f & 1 & 0 & 0\\0 & e & 1 & 0\\0 & 0 & d & 1\end{matrix}\right]$

In [26]:
#Diagonal
print "$"+sympy.latex(Matrix.create_diagonal_matrix(4, symbolic=True))+"$"

$\left[\begin{matrix}g & 0 & 0 & 0\\0 & h & 0 & 0\\0 & 0 & i & 0\\0 & 0 & 0 & j\end{matrix}\right]$


$\left[\begin{matrix}g & 0 & 0 & 0\\0 & h & 0 & 0\\0 & 0 & i & 0\\0 & 0 & 0 & j\end{matrix}\right]$

In [27]:
#Upper bidiagonals
print "\n".join(["$"+sympy.latex(m)+"$" for m in Matrix.create_upper_bidiagonal_matrix(4, symbolic=True)])

$\left[\begin{matrix}1 & k & 0 & 0\\0 & 1 & l & 0\\0 & 0 & 1 & m\\0 & 0 & 0 & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & n & 0\\0 & 0 & 1 & o\\0 & 0 & 0 & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & 0 & 1 & p\\0 & 0 & 0 & 1\end{matrix}\right]$


$\left[\begin{matrix}1 & k & 0 & 0\\0 & 1 & l & 0\\0 & 0 & 1 & m\\0 & 0 & 0 & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & n & 0\\0 & 0 & 1 & o\\0 & 0 & 0 & 1\end{matrix}\right]$
$\left[\begin{matrix}1 & 0 & 0 & 0\\0 & 1 & 0 & 0\\0 & 0 & 1 & p\\0 & 0 & 0 & 1\end{matrix}\right]$

In [0]:
tnp = Matrix.create_totally_nonpositive_matrix(4,symbolic=True)["product"]

In [0]:
tnp_inv = tnp.inv()

In [0]:
tnp_inv_diag = tnp_inv.get_diagonals()
print "\n".join(["$"+sympy.latex(d) + "$" for d in tnp_inv_diag])

In [5]:
tnp

Matrix([
[        -1 - (a*c*h*l*o + a*c*h*l*p + a*c*h*n*p + a*e*h*l*o + a*e*h*l*p + a*e*h*n*p + a*i*m + a*i*o + a*i*p + b*e*h*l*o + b*e*h*l*p + b*e*h*n*p + b*i*m + b*i*o + b*i*p + j)/(a*c*f*k*n*p - i*m - i*o - i*p),                             k*(-1 - (a*c*h*l*o + a*c*h*l*p + a*c*h*n*p + a*e*h*l*o + a*e*h*l*p + a*e*h*n*p + a*i*m + a*i*o + a*i*p + b*e*h*l*o + b*e*h*l*p + b*e*h*n*p + b*i*m + b*i*o + b*i*p + j)/(a*c*f*k*n*p - i*m - i*o - i*p)),                                                     k*n*(-1 - (a*c*h*l*o + a*c*h*l*p + a*c*h*n*p + a*e*h*l*o + a*e*h*l*p + a*e*h*n*p + a*i*m + a*i*o + a*i*p + b*e*h*l*o + b*e*h*l*p + b*e*h*n*p + b*i*m + b*i*o + b*i*p + j)/(a*c*f*k*n*p - i*m - i*o - i*p)),                                                                               k*n*p*(-1 - (a*c*h*l*o + a*c*h*l*p + a*c*h*n*p + a*e*h*l*o + a*e*h*l*p + a*e*h*n*p + a*i*m + a*i*o + a*i*p + b*e*h*l*o + b*e*h*l*p + b*e*h*n*p + b*i*m + b*i*o + b*i*p + j)/(a*c*f*k*n*p - i*m - i*o - i*p))],
[    f*(-1 - 

# Submatrices

In [0]:
m = Matrix.create_symmetric_matrix(4)

In [19]:
m.get_submatrices(rows=2, cols=2)

[Matrix([
 [a11, a12],
 [a12, a22]]), Matrix([
 [a11, a13],
 [a12, a23]]), Matrix([
 [a11, a14],
 [a12, a24]]), Matrix([
 [a12, a13],
 [a22, a23]]), Matrix([
 [a12, a14],
 [a22, a24]]), Matrix([
 [a13, a14],
 [a23, a24]]), Matrix([
 [a11, a12],
 [a13, a23]]), Matrix([
 [a11, a13],
 [a13, a33]]), Matrix([
 [a11, a14],
 [a13, a34]]), Matrix([
 [a12, a13],
 [a23, a33]]), Matrix([
 [a12, a14],
 [a23, a34]]), Matrix([
 [a13, a14],
 [a33, a34]]), Matrix([
 [a11, a12],
 [a14, a24]]), Matrix([
 [a11, a13],
 [a14, a34]]), Matrix([
 [a11, a14],
 [a14, a44]]), Matrix([
 [a12, a13],
 [a24, a34]]), Matrix([
 [a12, a14],
 [a24, a44]]), Matrix([
 [a13, a14],
 [a34, a44]]), Matrix([
 [a12, a22],
 [a13, a23]]), Matrix([
 [a12, a23],
 [a13, a33]]), Matrix([
 [a12, a24],
 [a13, a34]]), Matrix([
 [a22, a23],
 [a23, a33]]), Matrix([
 [a22, a24],
 [a23, a34]]), Matrix([
 [a23, a24],
 [a33, a34]]), Matrix([
 [a12, a22],
 [a14, a24]]), Matrix([
 [a12, a23],
 [a14, a34]]), Matrix([
 [a12, a24],
 [a14, a44]]), 

In [0]:
sm2 = m.get_submatrices(rows=2, cols=2, rows_and_cols_consecutive=True, row_initials=True)

In [21]:
sm2

[Matrix([
 [a11, a12],
 [a12, a22]]), Matrix([
 [a12, a13],
 [a22, a23]]), Matrix([
 [a13, a14],
 [a23, a24]])]

In [22]:
print "\n".join(["$" + sympy.latex(x) + "$" for x in sm2])

$\left[\begin{matrix}a_{11} & a_{12}\\a_{12} & a_{22}\end{matrix}\right]$
$\left[\begin{matrix}a_{12} & a_{13}\\a_{22} & a_{23}\end{matrix}\right]$
$\left[\begin{matrix}a_{13} & a_{14}\\a_{23} & a_{24}\end{matrix}\right]$


$\left[\begin{matrix}a_{11} & a_{12}\\a_{12} & a_{22}\end{matrix}\right]$
$\left[\begin{matrix}a_{12} & a_{13}\\a_{22} & a_{23}\end{matrix}\right]$
$\left[\begin{matrix}a_{13} & a_{14}\\a_{23} & a_{24}\end{matrix}\right]$